File: src/geg/crypto/binding.py
"""The voter's binding of a ballot to the credential it was cast with.

The attestation is already bound to the *voter*: ``validate_ballot`` requires
``att.pseudonym == env.pseudonym`` and ``att.vk == env.vk``, and the ballot's
Schnorr signature is verified against that same ``vk``. A credential cannot be
lifted onto someone else's ballot.

What the voter never committed to was **which** of their own credentials this
ballot was cast with. ``canonical_ballot_message`` covers
``(election_id, pseudonym, ciphertexts, zk_proof)`` — not ``weight`` and not
``nonce``. So whoever assembled the ballot/attestation pair chose those two
fields, and the nonce is what orders re-votes:

    key = (envelope.attestation.nonce, sequence_number)     # core/admission

Given a voter's ballots A (nonce 100), B (200) and C (300), an assembler that
attaches C's attestation to ballot A and drops C makes A = (300, seq 0) outrank
B = (200, seq 1). The voter's last choice is discarded in favour of their first,
using only credentials the voter legitimately obtained. On a backend that issues
the credentials itself, no theft is needed at all — it simply picks the nonce.

This module closes that: a second Schnorr signature, under the same voter key,
over **the ballot and the credential together**.

Binding the pair is the point. A signature over the credential alone travels
*with* the credential, so the assembler moves both onto a different ballot and
the attack is unchanged. Only a message covering the ballot digest fixes the
pairing.

Nothing here is a new primitive. It is the house ``Transcript`` — length-prefixed,
domain-separated, byte-identical in the SDK — so the browser building the ballot
and the keyper validating it derive the same preimage from the same rules. The
signature is verified against the ballot's existing ``vk``, so a keyper needs no
new curve, no secp256k1 recovery and no chain context to check it.
"""

from __future__ import annotations

from eth_utils import keccak

from geg.crypto import schnorr
from geg.crypto.params import BINDING_LABEL
from geg.crypto.points import g1_from_compressed
from geg.crypto.transcript import Transcript

# The scheme is bound as a scalar rather than by name so the two credential
# schemes can never be read as one another: a LEGACY credential and a V1 one
# over the same fields must not produce the same binding.
_SCHEME_CODES = {"ATTESTATION_V1": 1, "ATTESTATION_LEGACY": 2}


def _scheme_code(scheme) -> int:
    name = getattr(scheme, "value", scheme)
    code = _SCHEME_CODES.get(name)
    if code is None:
        raise ValueError(f"unknown attestation scheme: {name!r}")
    return code


def binding_message(
    *,
    election_id: bytes,
    pseudonym: bytes,
    vk_bytes: bytes,
    ballot_digest: bytes,
    attestation_scheme,
    attestation_digest: bytes,
    eligibility_signature: bytes,
) -> bytes:
    """The keccak256 digest the voter signs to bind one ballot to one credential.

    ``ballot_digest`` is ``keccak(canonical_ballot_message(...))`` and
    ``attestation_digest`` is the credential's own signed digest — ``V1`` and
    ``LEGACY`` each already define one, and reusing them keeps a single
    definition of each message rather than a second copy that can drift.

    ``eligibility_signature`` is included so the voter commits to the exact
    credential *instance*, not merely to its contents. It costs nothing and is
    strictly tighter.
    """
    if len(election_id) != 32:
        raise ValueError(f"electionId must be 32 bytes (got {len(election_id)})")
    if len(pseudonym) != 32:
        raise ValueError(f"pseudonym must be 32 bytes (got {len(pseudonym)})")
    if len(vk_bytes) != 48:
        raise ValueError(f"vk must be 48 bytes (got {len(vk_bytes)})")
    if len(ballot_digest) != 32:
        raise ValueError(f"ballotDigest must be 32 bytes (got {len(ballot_digest)})")
    if len(attestation_digest) != 32:
        raise ValueError(f"attestationDigest must be 32 bytes (got {len(attestation_digest)})")
    if not eligibility_signature:
        raise ValueError("eligibilitySignature must not be empty")

    t = Transcript(BINDING_LABEL)
    t.append("bind:electionId", election_id)
    t.append("bind:pseudonym", pseudonym)
    t.append("bind:vk", vk_bytes)
    t.append("bind:ballot", ballot_digest)
    t.append_scalar("bind:scheme", _scheme_code(attestation_scheme))
    t.append("bind:attestation", attestation_digest)
    t.append("bind:eligSig", eligibility_signature)
    return keccak(t.preimage())


def attestation_digest_for(att) -> bytes:
    """The credential's own signed digest, whichever scheme it uses."""
    from geg.crypto.attestation import attestation_message, legacy_attestation_message
    from geg.envelopes.types import AttestationScheme

    if att.scheme is AttestationScheme.V1:
        return attestation_message(att.election_id, att.pseudonym, att.vk, att.weight, att.nonce)
    if att.scheme is AttestationScheme.LEGACY:
        return legacy_attestation_message(att.election_id, att.pseudonym, att.vk)
    raise ValueError(f"unknown attestation scheme: {att.scheme!r}")


def ballot_message_digest(env) -> bytes:
    """``keccak(canonical_ballot_message(...))`` for an envelope.

    The same preimage the voter's *ballot* signature is taken over, so the binding
    inherits its coverage — election, pseudonym, every ciphertext and the proof —
    without restating it and risking a second, divergent definition.
    """
    from geg.crypto.ballot import canonical_ballot_message

    return keccak(
        canonical_ballot_message(
            env.election_id,
            env.pseudonym,
            [(ct.c1, ct.c2) for ct in env.ciphertexts],
            env.zk_proof,
        )
    )


def envelope_binding_message(env, ballot_digest: bytes) -> bytes:
    """``binding_message`` for a whole ballot envelope — the form callers want."""
    return binding_message(
        election_id=env.election_id,
        pseudonym=env.pseudonym,
        vk_bytes=env.vk,
        ballot_digest=ballot_digest,
        attestation_scheme=env.attestation.scheme,
        attestation_digest=attestation_digest_for(env.attestation),
        eligibility_signature=env.attestation.signature,
    )


def sign_ballot_binding(
    *,
    voter_sk: int,
    voter_vk,
    election_id: bytes,
    pseudonym: bytes,
    vk_bytes: bytes,
    ciphertexts,
    zk_proof: bytes,
    attestation,
    k: int | None = None,
) -> bytes:
    """Produce the binding signature from the parts a voter holds at build time.

    This is the whole signer-side API: a caller that has just built a ballot and
    obtained a credential should never assemble the message itself, because a
    second assembly is a second chance to disagree with the verifier about it.
    """
    from geg.crypto.ballot import canonical_ballot_message

    ballot_digest = keccak(
        canonical_ballot_message(election_id, pseudonym, list(ciphertexts), zk_proof)
    )
    message = binding_message(
        election_id=election_id,
        pseudonym=pseudonym,
        vk_bytes=vk_bytes,
        ballot_digest=ballot_digest,
        attestation_scheme=attestation.scheme,
        attestation_digest=attestation_digest_for(attestation),
        eligibility_signature=attestation.signature,
    )
    return sign_binding(voter_sk, voter_vk, message, k=k)


def sign_binding(voter_sk: int, voter_vk, message: bytes, *, k: int | None = None) -> bytes:
    """Sign a binding message with the voter's ballot key (80-byte ``R ‖ s``)."""
    R, s = schnorr.sign(voter_sk, voter_vk, message, k=k)
    return schnorr.encode(R, s)


def verify_binding_sig(vk_bytes: bytes, message: bytes, signature: bytes) -> bool:
    """Verify a binding signature. Returns ``False`` (never raises) on any
    malformed input, so callers treat a ``False`` uniformly as a rejection."""
    try:
        vk = g1_from_compressed(vk_bytes)
        R, s = schnorr.decode(signature)
        return schnorr.verify(vk, message, R, s)
    except Exception:  # noqa: BLE001
        return False

File: frontend/shared/src/binding.ts
/** The voter's binding of a ballot to the credential it was cast with.
 *
 * The credential already names this voter — admission checks
 * `att.pseudonym === ballot.pseudonym` and `att.vk === ballot.vk`, and the ballot
 * signature is verified against that same `vk`. What nothing voter-signed covered
 * is *which* of that voter's credentials this ballot was cast with:
 * `canonicalBallotMessage` spans `(electionId, pseudonym, ciphertexts, zkProof)`
 * and neither `weight` nor `nonce`. Since `nonce` orders re-votes, whoever pairs a
 * ballot with a credential could otherwise choose which of a voter's ballots wins.
 *
 * Byte-identical to `geg/crypto/binding.py`, and pinned against vectors generated
 * by it. If the two ever disagree, every ballot this app produces is excluded as
 * INVALID_ATTESTATION — a failure that reads as an eligibility problem and gives
 * no hint that it is really an encoding one.
 *
 * ## Why the transcript is rebuilt here
 *
 * The SDK exports `Transcript`, but only its `challenge()` — `preimage()` and the
 * accumulated parts are private, so it cannot produce the bytes this message
 * hashes. The layout is reproduced below rather than worked around: it is four
 * lines, it is fully specified, and the cross-language vectors are what actually
 * hold it in place.
 */

import { keccak256 } from "viem";
import type { Hex } from "viem";
import { hexToBytes } from "./hex";
import type { AttestationJson } from "./types";

/** Transcript label; distinct from the ballot and attestation labels so a
 * signature over one can never be presented as a signature over another. */
const BINDING_LABEL = "SHUTTER-VOTE-BINDING-v1";
const ATTESTATION_LABEL = "SHUTTER-VOTE-ATTEST-v1";

/** Credential schemes, as scalars. Bound explicitly so a LEGACY credential and a
 * V1 one over otherwise identical fields cannot produce the same binding. */
const SCHEME_CODES: Record<string, bigint> = {
  ATTESTATION_V1: 1n,
  ATTESTATION_LEGACY: 2n,
};

const enc = new TextEncoder();

function u32be(n: number): Uint8Array {
  return new Uint8Array([(n >>> 24) & 0xff, (n >>> 16) & 0xff, (n >>> 8) & 0xff, n & 0xff]);
}

/** 32-byte big-endian, matching `params.scalar_to_bytes`. */
function scalarToBytes(s: bigint): Uint8Array {
  const out = new Uint8Array(32);
  let v = s;
  for (let i = 31; i >= 0 && v > 0n; i--) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

/** The append-only, length-prefixed transcript of `crypto/transcript.py`.
 * Length prefixes are what make concatenation injective — without them
 * `("ab", "c")` and `("a", "bc")` would hash alike. */
class Preimage {
  private readonly parts: Uint8Array[];

  constructor(label: string) {
    this.parts = [enc.encode(label)];
  }

  append(tag: string, value: Uint8Array): this {
    const tb = enc.encode(tag);
    this.parts.push(u32be(tb.length), tb, u32be(value.length), value);
    return this;
  }

  appendScalar(tag: string, s: bigint): this {
    return this.append(tag, scalarToBytes(s));
  }

  bytes(): Uint8Array {
    const total = this.parts.reduce((n, p) => n + p.length, 0);
    const out = new Uint8Array(total);
    let at = 0;
    for (const p of this.parts) {
      out.set(p, at);
      at += p.length;
    }
    return out;
  }
}

function requireLen(name: string, b: Uint8Array, n: number): Uint8Array {
  if (b.length !== n) throw new Error(`${name} must be ${n} bytes (got ${b.length})`);
  return b;
}

/** The credential's own signed digest — the same one the eligibility service
 * signed, not a second definition of it. */
export function attestationDigest(att: AttestationJson): Uint8Array {
  const electionId = requireLen("electionId", hexToBytes(att.electionId), 32);
  const pseudonym = requireLen("pseudonym", hexToBytes(att.pseudonym), 32);
  const vk = requireLen("vk", hexToBytes(att.vk), 48);
  const scheme = att.scheme ?? "ATTESTATION_V1";

  if (scheme === "ATTESTATION_LEGACY") {
    // Weightless: a bare keccak concatenation with no domain separator, kept for
    // interop. Deliberately *not* the V1 transcript.
    const flat = new Uint8Array(32 + 32 + 48);
    flat.set(electionId, 0);
    flat.set(pseudonym, 32);
    flat.set(vk, 64);
    return hexToBytes(keccak256(flat));
  }

  const nonce = att.nonce ?? 0;
  const t = new Preimage(ATTESTATION_LABEL)
    .append("attest:electionId", electionId)
    .append("attest:pseudonym", pseudonym)
    .append("attest:vk", vk)
    .appendScalar("attest:weight", BigInt(att.weight))
    .appendScalar("attest:nonce", BigInt(nonce));
  return hexToBytes(keccak256(t.bytes()));
}

export interface BindingMessageArgs {
  electionId: Uint8Array;
  pseudonym: Uint8Array;
  vk: Uint8Array;
  /** `keccak256(canonicalBallotMessage(...))` — the SDK builds the preimage. */
  ballotDigest: Uint8Array;
  attestation: AttestationJson;
}

/** The keccak256 digest the voter signs to bind one ballot to one credential. */
export function bindingMessage(args: BindingMessageArgs): Uint8Array {
  const scheme = args.attestation.scheme ?? "ATTESTATION_V1";
  const code = SCHEME_CODES[scheme];
  if (code === undefined) throw new Error(`unknown attestation scheme: ${scheme}`);

  const eligSig = hexToBytes(args.attestation.signature);
  if (eligSig.length === 0) throw new Error("eligibilitySignature must not be empty");

  const t = new Preimage(BINDING_LABEL)
    .append("bind:electionId", requireLen("electionId", args.electionId, 32))
    .append("bind:pseudonym", requireLen("pseudonym", args.pseudonym, 32))
    .append("bind:vk", requireLen("vk", args.vk, 48))
    .append("bind:ballot", requireLen("ballotDigest", args.ballotDigest, 32))
    .appendScalar("bind:scheme", code)
    // The credential's contents *and* the issuer's signature over them, so the
    // voter commits to the exact instance rather than to fields that a second,
    // differently-signed credential could also satisfy.
    .append("bind:attestation", attestationDigest(args.attestation))
    .append("bind:eligSig", eligSig);
  return hexToBytes(keccak256(t.bytes()));
}

/** Convenience for callers holding hex rather than bytes. */
export function bindingMessageHex(args: {
  electionId: Hex;
  pseudonym: Hex;
  vk: Hex;
  ballotDigest: Hex;
  attestation: AttestationJson;
}): Uint8Array {
  return bindingMessage({
    electionId: hexToBytes(args.electionId),
    pseudonym: hexToBytes(args.pseudonym),
    vk: hexToBytes(args.vk),
    ballotDigest: hexToBytes(args.ballotDigest),
    attestation: args.attestation,
  });
}
